src/main/scala/org/constellation/domain/redownload/RedownloadService.scala

```diff
@@ -1,38 +1,170 @@
+  private[redownload] val proposedSnapshots: Ref[F, Map[Id, Proposals]] = Ref.unsafe(Map.empty)
 
-  def persistOwnSnapshot(height: Long, hash: String): F[Unit] =
-    ownSnapshots.modify { m =>
-      val updated = if (m.get(height).nonEmpty) m else m.updated(height, hash)
-      (updated, ())
-    }
+  def persistLocalSnapshot(recentSnapshot: RecentSnapshot) = updateProposedSnapshots(Map((cluster.id, Map(recentSnapshot.height -> recentSnapshot))))
+
+  def getLocalSnapshots() = proposedSnapshots.get.map { snapshots =>
+    snapshots.getOrElse(cluster.id, Map())
+  }
+
+  def getLocalSnapshotAtHeight(height: Long): F[Option[RecentSnapshot]] = proposedSnapshots.get.map { proposedSnaps =>
+    val localSnapshots = proposedSnaps.getOrElse(cluster.id, Map())
+    localSnapshots.get(height)
+  }
 
-  def getOwnSnapshots(): F[Map[Long, String]] = ownSnapshots.get
+  private[redownload] def fetchPeerProposals() =
+    for {
+      peers <- cluster.readyPeers
+      chunkedPeerProposals <- peers.toList.traverse {
+        case (id, peerData) =>
+          val resp = peerData.client.getNonBlockingFLogged[F, Array[Array[Byte]]](
+            "snapshot/own",
+            timeout = 3.seconds,
+            tag = "snapshot/own"
+          )(C)
+          resp.map((id, _)).handleErrorWith { ex =>
+            F.delay {logger.error(s"fetchPeerProposals error - ${ex}") } >> F.pure((id, Array()))
+          }
+      }
+      deSerializedPeerProposals = chunkedPeerProposals.map(deserializeProposals).flatMap {
+        case (id, recentSnaps) =>
+          val heightRecentSnapMap = recentSnaps.map { recentSnap => (recentSnap.height, recentSnap)}.toMap
+          recentSnaps.map(snap => (id, heightRecentSnapMap))
+      }.toMap
+    } yield deSerializedPeerProposals
 
-  def getOwnSnapshot(height: Long): F[Option[String]] =
-    ownSnapshots.get.map(_.get(height))
+  private[redownload] def updateProposedSnapshots(fetchedProposals: Map[Id, Proposals]) =
+    proposedSnapshots.modify { m =>
+      val updatedProposals = fetchedProposals.foldLeft(m) {
+        case (prevProposals, (id, proposals)) =>
+          val proposalsRecievedFromId = prevProposals.get(id).fold(proposals) { prevProposals =>
+            val newProposals = proposals.filter { case (existingHeight, existingSnap) => prevProposals.get(existingHeight).isEmpty }
+            prevProposals ++ newProposals
+          }
+          prevProposals.updated(id, proposalsRecievedFromId)
+      }
+      (updatedProposals, ())
+    }
 
-  // TODO: Use cluster peers to fetch "/snapshot/own" and merge with peersProposals Map
-  def fetchPeersProposals(): F[Unit] = ???
+  def fetchAndSetPeerProposals() = fetchPeerProposals().flatMap(updateProposedSnapshots)
 
-  // TODO: Check for the majority snapshot and store under a variable
-  def recalculateMajoritySnapshot(): F[Unit] = ???
+  def recalculateMajoritySnapshot(): F[ReDownloadPlan] =
+    for {
+      allPeers <- cluster.getPeerInfo
```

Comment:
Why are you fetching proposals from ready peers but recalculating for all the peers? It can be that those peers are offline, leaving, etc.